Avoid copying FontPalette and FontVariantAlternates

brentfulgham opened the pull request:
#### 371cab863ee8cb0010261fddb8e0435391fc4dbd
<pre>
Avoid copying FontPalette and FontVariantAlternates
<a href="https://bugs.webkit.org/show_bug.cgi?id=273568">https://bugs.webkit.org/show_bug.cgi?id=273568</a>
&lt;<a href="https://rdar.apple.com/problem/127380343">rdar://problem/127380343</a>&gt;

Reviewed by Chris Dumez.

Reduce unnecessary copies in our font handling code:
(1) FontPalette is a struct containing an enum and an AtomString, and should be returned
    by const reference to avoid copying that string.
(2) FontVariantAlternates is lightweight, containing only a `Markable`, but the
    `Markable` is holding a very heavy-weight object that contains Vectors of Strings,
    and Strings.

* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::fontFace):
* Source/WebCore/css/CSSFontFaceSet.h:
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::variantAlternates const):
(WebCore::FontDescription::fontPalette const):
(WebCore::FontDescription::setVariantAlternates):
(WebCore::FontDescription::setFontPalette):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setFontPalette):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::fontPalette const):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates):

Canonical link: <a href="https://commits.webkit.org/278351@main">https://commits.webkit.org/278351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37c10343e447946c45427feaabd4fbc31132b9c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29528 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53494 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/925 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/551 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40990 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27197 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22088 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24617 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/491 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46604 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/553 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55079 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25331 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48401 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26592 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43425 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47426 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11031 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27456 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26324 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
It's "FontVariantAlternates" and "FontPalette" fwiw

cdumez commented on Source/WebCore/css/CSSFontFace.h:
```diff
@@ -107,7 +107,7 @@ class CSSFontFace final : public RefCounted<CSSFontFace> {
     std::span<const UnicodeRange> ranges() const { ASSERT(m_status != Status::Failure); return m_ranges.span(); }
 
-    void setFontSelectionCapabilities(FontSelectionCapabilities capabilities) { m_fontSelectionCapabilities = capabilities; }
+    void setFontSelectionCapabilities(const FontSelectionCapabilities& capabilities) { m_fontSelectionCapabilities = capabilities; }
```
This is not a large type, I am not convinced passing by const reference is actually better here?

brentfulgham replied:
Will remove.

cdumez commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -447,7 +447,7 @@ ExceptionOr<bool> CSSFontFaceSet::check(const String& font, const String& text)
 }
 
-CSSSegmentedFontFace* CSSFontFaceSet::fontFace(FontSelectionRequest request, const AtomString& family)
+CSSSegmentedFontFace* CSSFontFaceSet::fontFace(const FontSelectionRequest& request, const AtomString& family)
```
Not a large type, not obviously clear this is actually better?

brentfulgham replied:
Will revert.

cdumez commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -80,7 +80,7 @@ class FontDescription {
     FontVariantNumericOrdinal variantNumericOrdinal() const { return static_cast<FontVariantNumericOrdinal>(m_variantNumericOrdinal); }
     FontVariantNumericSlashedZero variantNumericSlashedZero() const { return static_cast<FontVariantNumericSlashedZero>(m_variantNumericSlashedZero); }
-    FontVariantAlternates variantAlternates() const { return m_variantAlternates; }
+    const FontVariantAlternates& variantAlternates() const { return m_variantAlternates; }
```
Not a large type, not clear this is actually better. Passing small types by value is often faster than by reference. This looks like less than 2 words.

brentfulgham replied:
This is a `Markable` that wraps a large object. I think this one does make sense to pass by reference.

cdumez replied:
Agreed, I had not looked correctly.

cdumez commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -108,7 +108,7 @@ class FontDescription {
     AllowUserInstalledFonts shouldAllowUserInstalledFonts() const { return static_cast<AllowUserInstalledFonts>(m_shouldAllowUserInstalledFonts); }
     bool shouldDisableLigaturesForSpacing() const { return m_shouldDisableLigaturesForSpacing; }
-    FontPalette fontPalette() const { return m_fontPalette; }
+    const FontPalette& fontPalette() const { return m_fontPalette; }
```
FontPalette is essentially an AtomString so this one makes total sense to me, to avoid ref-counting churn.

cdumez commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1635,7 +1635,7 @@ inline void BuilderCustom::applyInitialFontVariantAlternates(BuilderState& build
 inline void BuilderCustom::applyValueFontVariantAlternates(BuilderState& builderState, CSSValue& value)
 {
-    auto setAlternates = [&builderState](FontVariantAlternates alternates) {
+    auto setAlternates = [&builderState](FontVariantAlternates&& alternates) {
```
Given how it is used, I think I would have used `const FontVariantAlternates&` here..

brentfulgham replied:
Will do!

cdumez commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1635,7 +1635,7 @@ inline void BuilderCustom::applyInitialFontVariantAlternates(BuilderState& build
-    auto setAlternates = [&builderState](FontVariantAlternates alternates) {
+    auto setAlternates = [&builderState](FontVariantAlternates&& alternates) {
         auto fontDescription = builderState.fontDescription();
         fontDescription.setVariantAlternates(WTFMove(alternates));
```
.. dropped the `WTFMove()` here ..

cdumez commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1656,7 +1656,7 @@ inline void BuilderCustom::applyValueFontVariantAlternates(BuilderState& builder
 
     if (auto* alternatesValues = dynamicDowncast<CSSFontVariantAlternatesValue>(value)) {
-        setAlternates(alternatesValues->value());
+        setAlternates(FontVariantAlternates(alternatesValues->value()));
```
.. and kept this code as-is.

The perf is the same and it is a little simpler.

cdumez reviewed (approved):
LGTM with suggestion.

brentfulgham committed:
Avoid copying FontPalette and FontVariantAlternates
https://bugs.webkit.org/show_bug.cgi?id=273568
<rdar://problem/127380343>

Reviewed by Chris Dumez.

Reduce unnecessary copies in our font handling code:
(1) FontPalette is a struct containing an enum and an AtomString, and should be returned
    by const reference to avoid copying that string.
(2) FontVariantAlternates is lightweight, containing only a `Markable`, but the
    `Markable` is holding a very heavy-weight object that contains Vectors of Strings,
    and Strings.

* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::fontFace):
* Source/WebCore/css/CSSFontFaceSet.h:
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::variantAlternates const):
(WebCore::FontDescription::fontPalette const):
(WebCore::FontDescription::setVariantAlternates):
(WebCore::FontDescription::setFontPalette):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setFontPalette):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::fontPalette const):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates):

Canonical link: https://commits.webkit.org/278351@main
